docs/rfcs/sync-manager.md

```diff
@@ -0,0 +1,171 @@
+# Sync manager
+
+We've identified the need for a "sync manager" (although are yet to identify a good name for it!) This manager will be responsible for managing "global sync state" and coordinating each engine. For the purposes of this, we are including the "clients" engine in this global state.
+
+The primary responsibilities of the engine are:
+
+* Manage the "meta/global" resource. This resources holds the following information:
+    * syncID - the global syncID for the entire storage managed by sync.
+    * storageVersion - an integer. If this value is greater than a sync client knows about, that client will refuse to sync. While this is rarely changed, it allows us to lock out older sync clients which could be useful.
+    * The list of "declined" engines - that is, engines which the user has explicitly declined from syncing. Note that declined engines are global for the account - declining an engine on any device should decline it on every device. While this is something we'd like to consider changing in the future, for now we are sticking with this global declined state.
+    * Some metadata about every collection (which may be subtly different from an "engine"). Currently this metadata is only {version, syncID}
+
+    Management of meta/global includes all syncIDs - if there's no global syncID, or an engine has no syncID, it should generate one. As we mention below, engines will be responsible for persisting the syncID and version and taking necessary action when it changes. Thus, engines never need to communicate a new syncID back to the global.
+
+* Query info/collections and pass this information to engines.
+
+* Manage interaction with the token server to obtain a token. As part of this, the manager needs to detect when an engine fails to use the token, generally due to a node reassignment mid sync. In this scenario, the manager will abort any pending syncs and reset itself, starting a new sync flow. Note also that this means the containing application will generally not need to see authentication errors from engines, only from the manager.
+
+* Manage the "clients" collection - we probably can't ignore this any longer, especially for bookmarks (as desktop will send a wipe command on bookmark restore). This involves both communicating with the engine regarding commands targetting it, and accepting commands to be send to other devices. Note however that it's not yet clear where these commands will originate from - eg, consider a bookmark restore - which point in this process will queue the wipe commands for other clients?). Exactly how to manage this (eg, when we can assume the engine has processed the command?) are TBD.
```

Comment:
Ugh yeah, forgot about the fact that bookmarks is likely to need it for correctness.

Reply:
We can maybe punt on this for the Fenix beta (you'll just get all your old bookmarks back 🙃), but I think we'll need to handle client commands somehow before shipping. We talked about moving them over to push, but, given the need to support existing clients that restore, I think we'll need to have _something_ that handles them.

Reply:
> We talked about moving them over to push

TBH I'm not convinced push is a good replacement for engine commands. OTOH, i suspect wipes are the only commands with value, so an alternative for that might be possible (but one doesn't spring to mind)